runtime/cmd/decode-state-values/main.go

```diff
@@ -22,113 +22,362 @@ package main
-	Key   key
+func decodeSlab(id atree.StorageID, data []byte) (atree.Slab, error) {
+	return atree.DecodeSlab(
+		id,
+		data,
+		interpreter.CBORDecMode,
+		interpreter.DecodeStorable,
+		interpreter.DecodeTypeInfo,
+	)
 }
 
-func worker(jobs <-chan entry, wg *sync.WaitGroup, decoded *uint64) {
-	defer wg.Done()
+func storageIDStorageKey(id atree.StorageID) storageKey {
+	return storageKey{
+		string(id.Address[:]),
+		"",
+		"$" + string(id.Index[:]),
+	}
+}
 
-	var err error
-	var data []byte
+// slabStorage
 
-	for e := range jobs {
+type slabStorage struct{}
 
-		data, err = hex.DecodeString(e.Value)
-		if err != nil {
-			log.Fatal(err)
-		}
+var _ atree.SlabStorage = &slabStorage{}
+
+func (s *slabStorage) Retrieve(id atree.StorageID) (atree.Slab, bool, error) {
+	data, ok := storage[storageIDStorageKey(id)]
+	if !ok {
+		return nil, false, nil
+	}
+
+	slab, err := decodeSlab(id, data)
+	if err != nil {
+		return nil, true, err
+	}
+
+	return slab, true, nil
+}
+
+func (s *slabStorage) Store(_ atree.StorageID, _ atree.Slab) error {
+	panic("unexpected Store call")
+}
 
-		var version uint16
-		data, version = interpreter.StripMagic(data)
-		if version == 0 {
+func (s *slabStorage) Remove(_ atree.StorageID) error {
+	panic("unexpected Remove call")
+}
+
+func (s *slabStorage) GenerateStorageID(_ atree.Address) (atree.StorageID, error) {
+	panic("unexpected GenerateStorageID call")
+}
+
+func (s *slabStorage) SlabIterator() (atree.SlabIterator, error) {
+	var slabs []struct {
+		atree.StorageID
+		storageKey
+	}
+
+	// NOTE: iteration over map is safe,
+	// as result is sorted below
+
+	for key := range storage { //nolint:maprangecheck
+
+		var address atree.Address
+		copy(address[:], key[0])
+		storageID := storageKeySlabStorageID(address, key[2])
+		if storageID == atree.StorageIDUndefined {
 			continue
 		}
 
-		rawOwner, err := hex.DecodeString(e.Key.KeyParts[1].Value)
-		if err != nil {
-			log.Fatal(err)
-		}
+		slabs = append(slabs, struct {
+			atree.StorageID
+			storageKey
+		}{
+			StorageID:  storageID,
+			storageKey: key,
+		})
+	}
 
-		owner := common.BytesToAddress(rawOwner)
+	sort.Slice(slabs, func(i, j int) bool {
+		a := slabs[i]
+		b := slabs[j]
+		return a.StorageID.Compare(b.StorageID) < 0
+	})
 
-		decodeFunction := interpreter.DecodeValue
-		if version <= 4 {
-			decodeFunction = interpreter.DecodeValueV4
+	var i int
+
+	bar := progressbar.Default(int64(len(slabs)))
+
+	return func() (atree.StorageID, atree.Slab) {
+		if i >= len(slabs) {
+			_ = bar.Close()
+			return atree.StorageIDUndefined, nil
 		}
 
-		var value interpreter.Value
-		value, err = decodeFunction(data, &owner, nil, version, nil)
+		slabEntry := slabs[i]
+		i++
+
+		_ = bar.Add(1)
+
+		storageID := slabEntry.StorageID
+		data := storage[slabEntry.storageKey]
+
+		slab, err := decodeSlab(storageID, data)
 		if err != nil {
-			log.Fatalf("failed to decode value: %s\n%s\n", err, e.Value)
+			log.Fatalf("failed to decode slab @ %s", storageID)
```

Comment:
we could clear the progress-bar before logging errors, i.e: `bar.Clear()`, so the bar will always stays below the rest of the logs.